JPA Map Storage with CRDB: OIDC Broker test failures

### Describe the bug

Several tests in the **org.keycloak.testsuite.broker** package present failures/errors with `CockroachDB`:

```
org.keycloak.testsuite.broker.AccountLinkTest#testAccountLink
org.keycloak.testsuite.broker.AccountLinkTest#testDeleteIdentityOnProviderRemoval
org.keycloak.testsuite.broker.KcOidcBrokerTest#testPostBrokerLoginFlowWithOTP
```

- Required JPA stores to reproduce the failures: **userSessions**

Logs contain a bunch of `Read timeout` messages, followed by

```
Caused by: org.postgresql.util.PSQLException: ERROR: restart transaction: TransactionRetryWithProtoRefreshError: TransactionAbortedError(ABORT_REASON_PUSHER_ABORTED): "sql txn" meta={id=3578b139 key=/Table/2262/1/"\xafѲ\xb4\x8e\xabC\xbb\xaa\xfe\xad\xf4\x86\xb8R\xc4"/0 pri=0.02463944 epo=0 ts=1657919486.368694475,0 min=1657919481.359664805,0 seq=35} lock=true stat=ABORTED rts=1657919481.359664805,0 wto=false gul=1657919481.859664805,0
  Hint: See: https://www.cockroachlabs.com/docs/v21.2/transaction-retry-error-reference.html#abort_reason_pusher_aborted
```

Apart from the above mentioned failures, that can be reproduced by simply configuring `userSessions` to use JPA storage, several other tests in this package also fail when other stores are also set to JPA:

```
./tests/base/target/surefire-reports/org.keycloak.testsuite.broker.AccountLinkTest.txt
./tests/base/target/surefire-reports/org.keycloak.testsuite.broker.ExternalKeycloakRoleToRoleMapperTest.txt
./tests/base/target/surefire-reports/org.keycloak.testsuite.broker.HardcodedRoleMapperTest.txt
./tests/base/target/surefire-reports/org.keycloak.testsuite.broker.HardcodedUserAttributeMapperTest.txt
./tests/base/target/surefire-reports/org.keycloak.testsuite.broker.JsonUserAttributeMapperTest.txt
./tests/base/target/surefire-reports/org.keycloak.testsuite.broker.KcOIDCBrokerWithSignatureTest.txt
./tests/base/target/surefire-reports/org.keycloak.testsuite.broker.KcOidcBrokerAcrParameterTest.txt
./tests/base/target/surefire-reports/org.keycloak.testsuite.broker.KcOidcBrokerClientSecretBasicAuthTest.txt
./tests/base/target/surefire-reports/org.keycloak.testsuite.broker.KcOidcBrokerClientSecretJwtCustomSignAlgTest.txt
./tests/base/target/surefire-reports/org.keycloak.testsuite.broker.KcOidcBrokerClientSecretJwtTest.txt
./tests/base/target/surefire-reports/org.keycloak.testsuite.broker.KcOidcBrokerEventTest.txt
./tests/base/target/surefire-reports/org.keycloak.testsuite.broker.KcOidcBrokerFrontendUrlTest.txt
./tests/base/target/surefire-reports/org.keycloak.testsuite.broker.KcOidcBrokerLoginHintTest.txt
./tests/base/target/surefire-reports/org.keycloak.testsuite.broker.KcOidcBrokerLogoutTest.txt
./tests/base/target/surefire-reports/org.keycloak.testsuite.broker.KcOidcBrokerNoLoginHintTest.txt
./tests/base/target/surefire-reports/org.keycloak.testsuite.broker.KcOidcBrokerNonceParameterTest.txt
./tests/base/target/surefire-reports/org.keycloak.testsuite.broker.KcOidcBrokerParameterForwardTest.txt
./tests/base/target/surefire-reports/org.keycloak.testsuite.broker.KcOidcBrokerPkceTest.txt
./tests/base/target/surefire-reports/org.keycloak.testsuite.broker.KcOidcBrokerPrivateKeyJwtCustomSignAlgTest.txt
./tests/base/target/surefire-reports/org.keycloak.testsuite.broker.KcOidcBrokerPrivateKeyJwtTest.txt
./tests/base/target/surefire-reports/org.keycloak.testsuite.broker.KcOidcBrokerPromptParameterTest.txt
./tests/base/target/surefire-reports/org.keycloak.testsuite.broker.KcOidcBrokerTest.txt
./tests/base/target/surefire-reports/org.keycloak.testsuite.broker.KcOidcBrokerUiLocalesDisabledTest.txt
./tests/base/target/surefire-reports/org.keycloak.testsuite.broker.KcOidcBrokerUiLocalesEnabledTest.txt
./tests/base/target/surefire-reports/org.keycloak.testsuite.broker.KcOidcUsernameTemplateMapperTest.txt
./tests/base/target/surefire-reports/org.keycloak.testsuite.broker.OidcAdvancedClaimToGroupMapperTest.txt
./tests/base/target/surefire-reports/org.keycloak.testsuite.broker.OidcAdvancedClaimToRoleMapperTest.txt
./tests/base/target/surefire-reports/org.keycloak.testsuite.broker.OidcClaimToRoleMapperTest.txt
./tests/base/target/surefire-reports/org.keycloak.testsuite.broker.OidcMultipleClaimToRoleMappersTest.txt
./tests/base/target/surefire-reports/org.keycloak.testsuite.broker.OidcUserAttributeMapperTest.txt
./tests/base/target/surefire-reports/org.keycloak.testsuite.broker.OidcUserInfoClaimToRoleMapperTest.txt
./tests/base/target/surefire-reports/org.keycloak.testsuite.broker.UsernameTemplateMapperTest.txt
```

- Required JPA storages to reproduce the failures: **authenticationSessions** and **userSessions**

The logs display several `Read timeout` messages and one of `CockroachDB`'s transaction retry errors:

```
Caused by: org.postgresql.util.PSQLException: ERROR: restart transaction: TransactionRetryWithProtoRefreshError: TransactionRetryError: retry txn (RETRY_SERIALIZABLE - failed preemptive refresh): "sql txn" meta={id=cfd2728d key=/Table/3110/1/"\n\x8f\xfe\xbe\a\x87I \x910\x83\xde2\x8a:\b"/0 pri=0.00013942 epo=0 ts=1658253792.299503598,2 min=1658253790.240125505,0 seq=45} lock=true stat=PENDING rts=1658253790.240125505,0 wto=false gul=1658253790.740125505,0
  Hint: See: https://www.cockroachlabs.com/docs/v21.2/transaction-retry-error-reference.html#retry_serializable
```
or

```
Caused by: org.postgresql.util.PSQLException: ERROR: restart transaction: TransactionRetryWithProtoRefreshError: WriteTooOldError: write at timestamp 1658239988.112544791,0 too old; wrote at 1658239988.156076839,1: "sql txn" meta={id=b9834d67 pri=0.05936828 epo=0 ts=1658239988.156076839,1 min=1658239988.112544791,0 seq=3} lock=true stat=PENDING rts=1658239988.112544791,0 wto=false gul=1658239988.612544791,0
  Hint: See: https://www.cockroachlabs.com/docs/v21.2/transaction-retry-error-reference.html
```

or even

```
Caused by: org.postgresql.util.PSQLException: ERROR: restart transaction: TransactionRetryWithProtoRefreshError: TransactionAbortedError(ABORT_REASON_PUSHER_ABORTED): "sql txn" meta={id=3578b139 key=/Table/2262/1/"\xafѲ\xb4\x8e\xabC\xbb\xaa\xfe\xad\xf4\x86\xb8R\xc4"/0 pri=0.02463944 epo=0 ts=1657919486.368694475,0 min=1657919481.359664805,0 seq=35} lock=true stat=ABORTED rts=1657919481.359664805,0 wto=false gul=1657919481.859664805,0
  Hint: See: https://www.cockroachlabs.com/docs/v21.2/transaction-retry-error-reference.html#abort_reason_pusher_aborted

```

**IMPORTANT**  When only `userSessions` are configured to use JPA (so `authenticationSessions` use, for example, `concurrenthashmap`, almost all of the broker tests pass, but even those that pass display error messages in the console during broker logout. For example:

```
10:47:36,583 WARN  [org.keycloak.broker.oidc.OIDCIdentityProvider] Failed backchannel broker logout to: https://localhost:8543/auth/realms/provider/protocol/openid-connect/logout?state=59482d04-1f2f-4310-8a74-84514e9d569a&id_token_hint=eyJhbGciOiJSUzI1NiIsInR5cCIgOiAiSldUIiwia2lkIiA6ICJzenljTXhOdFRvcHhuRExQZG5pXzlvNG9ZR2pwLWZrZ1c0a0dYU0FjaFJFIn0.eyJleHAiOjE2NTgyMzg3NTEsImlhdCI6MTY1ODIzODQ1MSwiYXV0aF90aW1lIjoxNjU4MjM4NDUxLCJqdGkiOiIxOTAxYzFkNi0zYTAwLTQxOTEtOTQ4MC05ZTEyY2ExZDJlZWMiLCJpc3MiOiJodHRwczovL2xvY2FsaG9zdDo4NTQzL2F1dGgvcmVhbG1zL3Byb3ZpZGVyIiwiYXVkIjoiYnJva2VyYXBwIiwic3ViIjoiNzJiMGJiZTUtNzI2Yi00YjVjLWJkOTctY2YxMmU5YjFjZjExIiwidHlwIjoiSUQiLCJhenAiOiJicm9rZXJhcHAiLCJub25jZSI6IkhoQmcxMmZkMWhsTjZJRnQ1R2JveFEiLCJzZXNzaW9uX3N0YXRlIjoiZjFiMjJjYzMtMDEwMC00Yzg5LTkwMTktMjUxZjU2NDhjOGQzIiwiYXRfaGFzaCI6IjV2LTVLdUdoVEUwNnF6bmwzeFNtS0EiLCJhY3IiOiIxIiwic2lkIjoiZjFiMjJjYzMtMDEwMC00Yzg5LTkwMTktMjUxZjU2NDhjOGQzIiwidXNlci1jbGFpbSI6eyJ0ZXN0IjoidmFsdWUifSwiZW1haWxfdmVyaWZpZWQiOnRydWUsInByZWZlcnJlZF91c2VybmFtZSI6InRlc3R1c2VyIiwiZW1haWwiOiJ1c2VyQGxvY2FsaG9zdC5jb20ifQ.Rp4MNWFbtxmgRkcD7HA-dKbnRqYWH5SyltNvwLAHs7ItOIGQ4mMv20a-PGQWDYMNWJKsH1TAMrwili_AGQeYjNZj2xNZbKjj-9RsY0VwBNeJcUOAk1jhH7_rGLJPoq4Ghncxxw3WC6Xz3ZLWd-XqCNs-DYvspmOm8_iVefAAcsPB1lzzMBRg9mhIwzWpFOjnUyKtbxrTrzEOJyUB8u9xG-rqJMYHXf-xPt2aGuNCb2t89KufdONxdLH5l6yY-gAVPicSIxFy-Nzx6XKUmM4kSXiniz92iJd8H8P-2VCgqixBz5c3GHW6O6bNKBBpaVrYAMN8TfItfup9Hkaogcy2ng

10:47:36,607 WARN  [org.keycloak.services] KC-SERVICES0057: Logout for client 'brokerapp' failed
java.net.SocketTimeoutException: Read timed out
...
10:47:36,664 ERROR [org.keycloak.services.error.KeycloakErrorHandler] Uncaught server error
javax.persistence.OptimisticLockException: org.hibernate.exception.LockAcquisitionException: could not execute statement
	at org.hibernate.internal.ExceptionConverterImpl.wrapLockException(ExceptionConverterImpl.java:277)
	at org.hibernate.internal.ExceptionConverterImpl.convert(ExceptionConverterImpl.java:98)
	at org.hibernate.internal.ExceptionConverterImpl.convert(ExceptionConverterImpl.java:181)
	at org.hibernate.query.internal.AbstractProducedQuery.executeUpdate(AbstractProducedQuery.java:1654)
	at org.keycloak.models.map.storage.jpa.JpaMapKeycloakTransaction.delete(JpaMapKeycloakTransaction.java:230)
	at org.keycloak.models.map.userSession.MapUserSessionProvider.removeUserSession(MapUserSessionProvider.java:318)
	at org.keycloak.services.managers.AuthenticationManager.finishBrowserLogout(AuthenticationManager.java:720)
	at org.keycloak.services.managers.AuthenticationManager.browserLogout(AuthenticationManager.java:655)
	at org.keycloak.protocol.oidc.endpoints.LogoutEndpoint.initiateBrowserLogout(LogoutEndpoint.java:750)
	at org.keycloak.protocol.oidc.endpoints.LogoutEndpoint.doBrowserLogout(LogoutEndpoint.java:447)
	at org.keycloak.protocol.oidc.endpoints.LogoutEndpoint.logoutConfirmAction(LogoutEndpoint.java:376)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:566)
...
Caused by: org.postgresql.util.PSQLException: ERROR: restart transaction: TransactionRetryWithProtoRefreshError: TransactionAbortedError(ABORT_REASON_PUSHER_ABORTED): "sql txn" meta={id=7f2aaebb key=/Table/3142/1/"YH-\x04\x1f/C\x10\x8at\x84QN\x9dV\x9a"/0 pri=0.02808664 epo=0 ts=1658238456.566812859,0 min=1658238451.565041195,0 seq=38} lock=true stat=ABORTED rts=1658238451.565041195,0 wto=false gul=1658238452.065041195,0
  Hint: See: https://www.cockroachlabs.com/docs/v21.2/transaction-retry-error-reference.html#abort_reason_pusher_aborted
...
10:47:41,772 DEBUG [org.keycloak.testsuite.broker.ExternalKeycloakRoleToRoleMapperTest] [Retriable] Timed out waiting for login page
10:47:41,922 DEBUG [org.keycloak.testsuite.broker.ExternalKeycloakRoleToRoleMapperTest] Clicking social kc-oidc-idp
```

Looks like `AbstractBaseBrokerTest.assertLoginPage` has some logic where navigation to the login page is retried a few times, and despite the error above the navigation ends up succeeding and tests pass.

### Version

main

### Expected behavior

Tests should run without failures or errors



### Actual behavior

_No response_

### How to Reproduce?

1- Setup CockroachDB according to the instructions in https://github.com/keycloak/keycloak/discussions/9596
2- Run any of the failing tests with the correct parameters, enabling JPA storage for **authenticationSessions**/**userSessions** only:

```
cd testsuite/integration-arquillian

mvn clean install -Pmap-storage -Dpageload.timeout=90000 -Dkeycloak.authSession.map.storage.provider=jpa \
-Dkeycloak.userSession.map.storage.provider=jpa \
-Dkeycloak.map.storage.connectionsJpa.url=jdbc:postgresql://localhost:26257/keycloak \
-Dkeycloak.map.storage.connectionsJpa.user=keycloak -Dkeycloak.map.storage.connectionsJpa.password=pass \
-Dtest=org.keycloak.testsuite.broker.**
```

Or replace the `-Dtest=` with the actual test you want to test if you don't want to test the whole package.

### Anything else?

_No response_

--- a/services/src/main/java/org/keycloak/broker/oidc/OIDCIdentityProvider.java
+++ b/services/src/main/java/org/keycloak/broker/oidc/OIDCIdentityProvider.java
@@ -18,6 +18,7 @@ package org.keycloak.broker.oidc;
 
 import com.fasterxml.jackson.databind.JsonNode;
 
+import org.apache.http.client.HttpClient;
 import org.jboss.logging.Logger;
 import org.keycloak.OAuth2Constants;
 import org.keycloak.OAuthErrorException;
@@ -30,6 +31,7 @@ import org.keycloak.broker.provider.util.SimpleHttp;
 import org.keycloak.common.util.Base64Url;
 import org.keycloak.common.util.SecretGenerator;
 import org.keycloak.common.util.Time;
+import org.keycloak.connections.httpclient.HttpClientProvider;
 import org.keycloak.crypto.KeyWrapper;
 import org.keycloak.crypto.SignatureProvider;
 import org.keycloak.events.Details;
@@ -39,6 +41,7 @@ import org.keycloak.events.EventType;
 import org.keycloak.jose.jws.JWSInput;
 import org.keycloak.jose.jws.JWSInputException;
 import org.keycloak.keys.loader.PublicKeyStorageManager;
+import org.keycloak.models.AbstractKeycloakTransaction;
 import org.keycloak.models.ClientModel;
 import org.keycloak.models.FederatedIdentityModel;
 import org.keycloak.models.KeycloakSession;
@@ -134,16 +137,28 @@ public class OIDCIdentityProvider extends AbstractOAuth2IdentityProvider<OIDCIde
         UriBuilder logoutUri = UriBuilder.fromUri(getConfig().getLogoutUrl())
                 .queryParam("state", sessionId);
         logoutUri.queryParam("id_token_hint", idToken);
-        String url = logoutUri.build().toString();
-        try {
-            int status = SimpleHttp.doGet(url, session).asStatus();
-            boolean success = status >= 200 && status < 400;
-            if (!success) {
-                logger.warn("Failed backchannel broker logout to: " + url);
+
+        final String url = logoutUri.build().toString();
+        final HttpClient client = session.getProvider(HttpClientProvider.class).getHttpClient();
+        session.getTransactionManager().enlistAfterCompletion(new AbstractKeycloakTransaction() {
+            @Override
+            protected void commitImpl() {
+                try {
+                    int status = SimpleHttp.doGet(url, client).asStatus();
+                    boolean success = status >= 200 && status < 400;
+                    if (!success) {
+                        logger.warn("Failed backchannel broker logout to: " + url);
+                    }
+                } catch (Exception e) {
+                    logger.warn("Failed backchannel broker logout to: " + url, e);
+                }
             }
-        } catch (Exception e) {
-            logger.warn("Failed backchannel broker logout to: " + url, e);
-        }
+
+            @Override
+            protected void rollbackImpl() {
+                // no-op
+            }
+        });
     }
 
 

